src/parser/jobs/blm/modules/RotationWatchdog.tsx

```diff
@@ -0,0 +1,563 @@
+	private history: Cycle[] = []
+
+	private firstEvent = true
+	// counters for suggestions
+	private missedF4s = 0
+	private extraF1s = 0
+	private extraT3s = 0
+	private rotationsWithoutFire = 0
+	private manafontBeforeDespair = 0
+	private astralFiresMissingDespairs = 0
+	private thunder3Casts = 0
+	private primaryTargetId: number | null = null
+
+	protected init() {
+		this.addHook('cast', {by: 'player'}, this.onCast)
+		this.addHook('complete', this.onComplete)
+		this.addHook(BLM_GAUGE_EVENT, this.onGaugeEvent)
+	}
+
+	// Handle events coming from BLM's Gauge module
+	private onGaugeEvent(event: BLMGaugeEvent) {
+		// If we're beginning the fire phase of this cycle, note it and save some data
+		if (this.currentGaugeState.astralFire === 0 && event.astralFire > 0) {
+			this.currentRotation.inFirePhase = true
+			this.currentRotation.firePhaseStartMP = this.combatants.selected.resources.mp
+
+			// If we didn't enter fire phase with a normal gauge state of 3 UI/UH stacks, note it
+			if (this.currentRotation.gaugeStateBeforeFire.umbralIce !== AFUIBUFFMAXSTACK ||
+				this.currentRotation.gaugeStateBeforeFire.umbralHearts !== AFUIBUFFMAXSTACK) {
+				this.currentRotation.atypicalAFStart = true
+			}
+		}
+
+		// If we no longer have enochian, flag it for display
+		if (this.currentGaugeState.enochian && !event.enochian) {
+			this.currentRotation.errorCode = ERROR_CODES.DROPPED_ENOCHIAN
+		}
+
+		// Retrieve the GaugeState from the event
+		this.currentGaugeState.astralFire = event.astralFire
+		this.currentGaugeState.umbralIce = event.umbralIce
+		this.currentGaugeState.umbralHearts = event.umbralHearts
+		this.currentGaugeState.enochian = event.enochian
+
+		// If we're in fire phase, stop processing
+		if (this.currentRotation.inFirePhase) {
+			return
+		}
+
+		// If we're in ice phase, set the current gauge state into the pre-fire cache for later recording
+		// Still need by-value assignment here
+		this.currentRotation.gaugeStateBeforeFire = Object.assign(this.currentRotation.gaugeStateBeforeFire, this.currentGaugeState)
+	}
+
+	// Handle cast events and updated recording data accordingly
+	private onCast(event: CastEvent) {
+		const actionId = event.ability.guid
+
+		if (!this.primaryTargetId && event.targetID) {
+			this.primaryTargetId = event.targetID
+		}
+
+		// If this action is signifies the beginning of a new cycle, unless this is the first
+		// cast of the log, stop the current cycle, and begin a new one. If Transposing from ice
+		// to fire, keep this cycle going
+		if (CYCLE_ENDPOINTS.includes(actionId) && !this.firstEvent &&
+			!(actionId === ACTIONS.TRANSPOSE.id && this.currentGaugeState.umbralIce > 0)) {
+			this.startRecording(event)
+		}
+		// Note that we've recorded our first cast now
+		if (this.firstEvent) { this.firstEvent = false }
+
+		// Add actions other than auto-attacks to the rotation cast list
+		const action = getDataBy(ACTIONS, 'id', actionId) as TODO
+		if (action && !action.autoAttack) {
```

Comment:
might be worth inverting this conditional given there's >1 chunk of logic nested